### PR TITLE
PCHR-1862: Add additional manager actions

### PR DESCRIPTION
**Pull request by dhavalsharma**
Implemented additional actions for manager. 
Disallowed manager to edit leave details like absence type, dates, day types. 
Allowed manager to edit details like sickness documents, toil expiry date.

## Gif
manager-action

## Post review gif for error message and date pickers
manager-action-error

## Notes 

1. `statusBeforeEdit` is used because `requestStatuses` is indexed by name and not value. Indexing it by value would means usage of numbers like 1,2,3 etc.,
2. `requestStatusesArray` is array of statuses because filter requires array 
3. Reorganized toil cases to group under create, edit or respond.
4. It is not expected that manager will not edit the toil request to only change expiry date and then send it back to staff. The expiry date is always obtained from server whenever staff edits the request.

## Side effects 

1. expiry date was not aligned to label. added class in another [PR](https://github.com/compucorp/civihr-employee-portal-theme/pull/151) 
2. `Weekend` was not setting properly, added test cases for weekend, working and non working days
3. changed working date in work pattern data from `2016-01-01` to `2016-01-05` otherwise it was giving incorrect result from public holidays api.
4. Updated `org.civicrm.reqangular/src/common/modules/models-instances.js` to include `'common/modules/apis'` as require dependency. 
This happened because `'common.apis'` angular module dependency was added in [commit](https://github.com/civicrm/civihr/commit/ba69c3376451416419fb1a8a862817a3ce9358e1).  @deb1990 ps have a look at the fix.

### CODE REVIEW

**Comment by AkA84** [comment]
@dhavalsharma two things stand out from your gif

1. The date fields are not wide enough to fully contain the dates
2. The yellow band contains unformatted json objects

both things need to be fixed, they're not acceptable to ship from a UI/UX standpoint

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js [hunk 1]
```diff
@@ -45,6 +45,11 @@ define([
        * @return {Promise}
        */
       vm.calculateToilExpiryDate = function () {
+        //do not calculate expiry date if its already set by manager for this leave request
+        if (vm.isRole('owner') && vm.isMode('edit')) {
+          return $q.resolve(vm.request.toil_expiry_date);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js]: Three things

1. You are referring to the manager in this comment (`if its already set by manager for this leave request`) but in the code you are referring the owner?
2. If you need to skip the calculation, then i think it's better to not just call the method in the first place, instead of having some exception handling in it
3. There is a TODO for this method, is it still valid?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js [hunk 2]
```diff
@@ -106,6 +111,15 @@ define([
       };
 
       /**
+      * updates expiry date when user changes it on ui
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js]:
```js
/**
 *
 */

Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js]: Start the documentation with an upper case

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js [hunk 3]
```diff
@@ -144,6 +160,15 @@ define([
           });
       }
 
+      /**
+       * Initialize expiryDate on UI from server date for manager's edit action
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js]: I don't understand this documentation, what do you mean "for manager's edit action"

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js [hunk 4]
```diff
@@ -144,6 +160,15 @@ define([
...
+       */
+      function initExpiryDate () {
+        if (vm.isRole('manager')) {
+          vm.uiOptions.expiryDate = vm._convertDateFormatFromServer(vm.request.toil_expiry_date);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js]: what happens if the user is the owner? the expiry date does not get shown?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/toil-request-ctrl_test.js [hunk 5]
```diff
@@ -53,19 +47,30 @@
...
+        Contact = _Contact_;
 
         spyOn($log, 'debug');
         spyOn(AbsenceTypeAPI, 'all').and.callThrough();
         spyOn(AbsenceTypeAPI, 'calculateToilExpiryDate').and.callThrough();
         spyOn(TOILRequestInstance, 'init').and.callThrough();
+        spyOn(Contact, 'all').and.callFake(function () {
+          return $q.resolve(ContactAPIMock.mockedContacts());
+        });
+
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/toil-request-ctrl_test.js]: empty line

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/toil-request-ctrl_test.js [hunk 6]
```diff
@@ -225,9 +230,64 @@
...
+          initTestController(directiveOptions);
+        });
+
+        it('manager role is set', function () {
+          expect($ctrl.isRole('manager')).toBeTruthy();
+        });
+
+        it('expiry date is set on ui', function () {
+          expect($ctrl.uiOptions.expiryDate).toEqual(jasmine.any(Date));
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/toil-request-ctrl_test.js]: isn't there a more meaningful test to do? i mean `expiryDate` must be set to some specific date rather that *any* date, no?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/toil-request-ctrl_test.js [hunk 7]
```diff
@@ -225,9 +230,64 @@
...
+          var beforeExpiryDate;
+
+          beforeEach(function () {
+            beforeExpiryDate = $ctrl.request.toil_expiry_date;
+            $ctrl.uiOptions.expiryDate = new Date();
+            $ctrl.updateExpiryDate();
+          });
+
+          it('sets new expiry date', function () {
+            expect(beforeExpiryDate).not.toEqual($ctrl.request.toil_expiry_date);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/toil-request-ctrl_test.js]: `beforeExpiryDate` is confusing, call it `original` or `old`

Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/toil-request-ctrl_test.js]: also, please test that the new value of `$ctrl.request.toil_expiry_date` is what you expect

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html [hunk 8]
```diff
@@ -284,7 +285,7 @@ <h4 class="modal-title">
       </div>
     </div>
     <div
-      ng-show="$ctrl.isRole('manager')"
+      ng-show="$ctrl.isRole('manager') && $ctrl.statusBeforeEdit.label === 'Waiting Approval'"
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html]: Do not rely on labels for any of your logic, use values of ids

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 9]
```diff
@@ -449,6 +449,26 @@ define([
       };
 
       /**
+       * Converts given date to server format
+       *
+       * @param {Date} date
+       * @return {Date} converted to server format
+       */
+      this._convertDateFormatToServer = function (date) {
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: private methods

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 10]
```diff
@@ -37,7 +37,7 @@ define([
       this.requestDayTypes = [];
       this.selectedAbsenceType = {};
       this.period = {};
-      this.statusLabel = '';
+      this.statusBeforeEdit = {};
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: it is before user starts to edit and it does not track the status of status upon edit , hence referred to before edit

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 11]
```diff
@@ -1111,6 +1110,13 @@ define([
       function updateRequest() {
         var self = this;
 
+        if (self.isRole('manager')) {
+          //if manager has not changed the status then reset status
+          if (!self.request.status_id) {
+            self.request.status_id = self.statusBeforeEdit.value;
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: for manager we filter the statuses (in `setStatuses` call which removes status which a manager can change in `requestStatuses` collection) and status before edit is set to null in html as its mapped to the model 
```
            <div class="chr_custom-select chr_custom-select--full">
              <select
                name="statusSelect"
                ng-options="status.value as status.label for status in $ctrl.requestStatuses"
                ng-model="$ctrl.request.status_id">
                <option value="">- select -</option>
              </select>
            </div>
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js (hunk 1 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js]: 1. It is to skip the call to `AbsenceType.calculateToilExpiryDate` in case `toil_expiry_date` is already modified by the manager.
2. see above point 
3. yes, TODO is still valid

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js (hunk 4 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js]: from html 

```
                  <div class="col-xs-4">
                    <span ng-if="!$ctrl.isRole('manager')">{{$ctrl.request.toil_expiry_date | date: $ctrl.uiOptions.userDateFormat}}</span>
                    <div ng-if="$ctrl.isRole('manager')"
                         class="input-group">
                      <input
                        type="text"
                        class="form-control"
                        uib-datepicker-popup
                        datepicker-options="$ctrl.uiOptions.date.expiry.options"
                        show-button-bar="false"
                        ng-model="$ctrl.uiOptions.expiryDate"
                        ng-change="$ctrl.updateExpiryDate()"
                        is-open="$ctrl.uiOptions.date.expiry.show"/>
                      <span class="input-group-btn">
                        <button
                          type="button"
                          class="btn btn-default fa fa-calendar"
                          ng-click="$ctrl.uiOptions.date.expiry.show = true"></button>
                      </span>
                    </div>
                  </div>
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/toil-request-ctrl_test.js (hunk 6 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/toil-request-ctrl_test.js]: It is randomly picked from collection , hence not using specific date
```
var toilRequest = TOILRequestInstance.init(mockData.findBy('status_id', status));
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 12]
```diff
@@ -789,7 +789,16 @@ define([
...
+            var errorMessages = [];
+
+            _.each(errors, function (value, key) {
+              var errorMessage = _.isArray(value) ? value.join(', ') : JSON.stringify(value);
+
+              errorMessage = errorMessage + ' for key ' + key;
+              errorMessages.push(errorMessage);
+            });
+
+            this.error = errorMessages.join(', ');
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]:
```js
if (_.isObject(errors)) {
  this.error = _.map(errors, function (value, key) {
    // Why is this necessary? how many different formats can `value` have?
    var errorMessage = _.isArray(value) ? value.join(', ') : JSON.stringify(value);

    return errorMessage + ' for key ' + key;
  }).join(', ');
}

Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: I don't know how many different formats error can have and that was the reason for having a discussion on standardization of error handling. I have seen this and the use of error_message so far.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 13]
```diff
@@ -469,6 +469,22 @@ define([
...
       /**
+       * Flattens statuses from object to array of objects. This is used to
+       * populate the dropdown with array of statuses.
+       *
+       * @return {Array}
+       */
+      this.getStatuses = function () {
+        if (this.requestStatuses) {
+          return _.map(this.requestStatuses, function (status) {
+              return status;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: wrong indentation

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 14]
```diff
@@ -469,6 +469,22 @@ define([
       }
 
       /**
+       * Flattens statuses from object to array of objects. This is used to
+       * populate the dropdown with array of statuses.
+       *
+       * @return {Array}
+       */
+      this.getStatuses = function () {
+        if (this.requestStatuses) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: `requestStatuses` should be initialized right at the beginning of the controller, so you can be sure it's always defined

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 15]
```diff
@@ -469,6 +469,22 @@ define([
...
+          return _.map(this.requestStatuses, function (status) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: isn't this equivalent to simply
```js
Object.values(this.requestStatuses)
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 16]
```diff
@@ -469,6 +469,22 @@ define([
...
+              return status;
+            });
+        }
+
+        return [];
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: if you initialize `this.requestStatuses` immediately, you don't need this

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 17]
```diff
@@ -483,6 +499,20 @@ define([
       };
 
       /**
+       * Checks if given status is applicable to manager
+       *
+       * @param {String} managerContactId
+       * @return {Boolean}
+       */
+      this.filterForManager = function (status) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: generic name (filter for what?), use a more descriptive one, something like `isStatusAvailableForManager`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 18]
```diff
@@ -483,6 +499,20 @@ define([
       };
 
       /**
+       * Checks if given status is applicable to manager
+       *
+       * @param {String} managerContactId
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: this is the wrong parameter

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 19]
```diff
@@ -483,6 +499,20 @@ define([
       };
 
       /**
+       * Checks if given status is applicable to manager
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: not "applicable", doesn't make sense in this sense. you are checking if the status is available for the manager to use

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 20]
```diff
@@ -483,6 +499,20 @@ define([
...
+       *
+       * @param {String} managerContactId
+       * @return {Boolean}
+       */
+      this.filterForManager = function (status) {
+        if (status.name === 'admin_approved' || status.name === 'waiting_approval') {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: one-liners when possible, please

```js
return !(status.name === 'admin_approved' || status.name === 'waiting_approval');
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 21]
```diff
@@ -743,15 +773,23 @@ define([
...
-            listToReturn = listOfDayTypes.filter(function (day) {
-              return day.name === 'weekend';
-            });
           }
         } catch (e) {
           listToReturn = [];
         }
 
+        if (!listToReturn.length) {
+          try {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: why do we need try/catch blocks here? (same as above)

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 22]
```diff
@@ -743,15 +773,23 @@ define([
...
+            if (this.calendar.isWeekend(moment(date))) {
+              listToReturn = listOfDayTypes.filter(function (day) {
+                return day.name === 'weekend';
+              });
+            }
+          } catch (e) {
+            listToReturn = [];
+          }
+        }
+
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: I'm not sure if i get exactly what is going on here, but wouldn't all of that be equivalent to something like
```js
/**
 * Gets list of day types if its found to be weekend or non working in calendar
 *
 * @param {Date} date to Checks
 * @param {Array} listOfDayTypes array of day types
 * @return {Array} non-empty if found else empty array
 */
function getDayTypesFromDate(date, listOfDayTypes) {
  var nameFilter = null;

  if (this.calendar.isNonWorkingDay(moment(date))) {
    nameFilter = 'non_working_day';
  } else if (this.calendar.isWeekend(moment(date))) {
    nameFilter = 'weekend';
  }

  return !nameFilter ? [] : listOfDayTypes.filter(function (day) {
    return day.name === nameFilter;
  });
}
```

**Comment by AkA84** [comment]
@dhavalsharma approved, but @deb1990 needs to approve that fix

**Comment by AkA84** [comment]
also @dhavalsharma please squash all those "PR reviews" commits into one
